### PR TITLE
Initial Classes

### PR DESCRIPTION
**Pull request by XedinUnknown**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -12,7 +12,12 @@
     "minimum-stability": "dev",
     "prefer-stable": true,
     "require": {
-        "php": "^5.4 | ^7.0"
+        "php": "^5.4 | ^7.0",
+        "dhii/output-renderer-interface": "^0.2",
+        "dhii/output-renderer-abstract": "dev-release/0.1",
+        "dhii/i18n-helper-base": "dev-develop",
+        "dhii/exception-helper-base": "dev-develop",
+        "dhii/stringable-interface": "^0.1"
```
Comment (mecha) [inline composer.json]: Dependency `dhii/stringable-interface` is not used for any invocation - only for type hinting in doc blocks.

Solution: make it a dev dependency.

File: src/AbstractBaseTemplate.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+use Exception as RootException;
+use Dhii\Output\Exception\RendererException;
+use Dhii\Output\Exception\TemplateRenderException;
+use Dhii\Util\String\StringableInterface as Stringable;
+use Dhii\I18n\StringTranslatingTrait;
+use Dhii\Exception\CreateInvalidArgumentExceptionCapableTrait;
+
+/**
+ * Base concrete functionality for blocks.
```
Comment (mecha) [inline src/AbstractBaseTemplate.php]: Description looks like an elusive copy-pasta rogue.

> Base concrete functionality for templates.

File: src/Exception/TemplateRenderException.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+use Dhii\Util\String\StringableInterface as Stringable;
+use Psr\Container\ContainerInterface;
+use Dhii\Output\TemplateInterface;
+use Dhii\Output\RendererAwareTrait;
+use Dhii\Output\ContextAwareTrait;
+use Dhii\I18n\StringTranslatingTrait;
+use Dhii\Exception\CreateInvalidArgumentExceptionCapableTrait;
+
+/**
+ * An exception that occurs when a renderer cannot produce output.
```
Comment (mecha) [inline src/Exception/TemplateRenderException.php]: Description looks like an elusive copy-pasta rogue.

> An exception that occurs in relation to a template.

File: test/unit/AbstractBaseBlockTest.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        $result = $_subject->_createRendererException($message, $code, $innerException);
+
+        $this->assertInstanceOf('Dhii\Output\Exception\RendererExceptionInterface', $result,
+            'The created message does not implement required interface');
+        $this->assertEquals($message, $result->getMessage(), 'The result message is wrong');
+        $this->assertEquals($code, $result->getCode(), 'The result code is wrong');
+        $this->assertEquals($innerException, $result->getPrevious(), 'The result inner exception is wrong');
+        $this->assertEquals($subject, $result->getRenderer(), 'The result renderer is wrong');
+    }
+}
```
Comment (mecha) [inline test/unit/AbstractBaseBlockTest.php]: Test is incomplete:

The `_createCouldNotRenderException()`, `_renderOnException()` and `_render()` methods are not tested.

File: test/unit/AbstractBaseTemplateTest.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        $result = $_subject->_createRendererException($message, $code, $innerException);
+
+        $this->assertInstanceOf('Dhii\Output\Exception\RendererExceptionInterface', $result,
+            'The created message does not implement required interface');
+        $this->assertEquals($message, $result->getMessage(), 'The result message is wrong');
+        $this->assertEquals($code, $result->getCode(), 'The result code is wrong');
+        $this->assertEquals($innerException, $result->getPrevious(), 'The result inner exception is wrong');
+        $this->assertEquals($subject, $result->getRenderer(), 'The result renderer is wrong');
+    }
+}
```
Comment (mecha) [inline test/unit/AbstractBaseTemplateTest.php]: Test is incomplete:

The `_createTemplateException()` and `_render()` methods are not tested.

File: test/functional/Exception/CouldNotRenderExceptionTest.php [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+     *
+     * @since [*next-version*]
+     *
+     * @param string        $message
+     * @param int           $code
+     * @param RootException $previous
+     *
+     * @return RootException The new exception.
+     */
+    public function createException($message = null, $code = null, RootException $previous = null)
```
Comment (mecha) [inline test/functional/Exception/CouldNotRenderExceptionTest.php:64]: Here and in every other test, I would prefer if the order of methods in test classes to be as follows, without any mixing:

* utility and factory methods first
* test methods second

**Review by mecha — CHANGES_REQUESTED** [review]

File: src/Exception/TemplateRenderException.php (hunk 3 above)
Comment (XedinUnknown) [reply src/Exception/TemplateRenderException.php]: Not quite the right alternative, but point taken.

File: test/unit/AbstractBaseBlockTest.php (hunk 4 above)
Comment (XedinUnknown) [reply test/unit/AbstractBaseBlockTest.php]: There's no `_render()` method; but there is a `render()`.

File: test/functional/Exception/CouldNotRenderExceptionTest.php (hunk 6 above)
Comment (XedinUnknown) [reply test/functional/Exception/CouldNotRenderExceptionTest.php:64]: Agreed. Can we do this later? This is kind of a pressing thing, and implementing your requested changes already set me back an hour and a half.

Comment (mecha) [reply test/functional/Exception/CouldNotRenderExceptionTest.php:64]: Sure. This is just for code tidiness more than a blocker.

**Review by mecha — APPROVED** [review]
LGTM

**Review by mecha — APPROVED** [review]
LGTM.

Travis build seems to have failed due to a problem I faced yesterday. Solution is to allow `dhii/output-renderer-interface` to be installed at `^0.2 | ^0.3`,

**Comment by mecha** [comment]
@XedinUnknown Travis build is successful. Kindly merge 😘

**Pull request merged**